SopraGameAction.cpp

```diff
@@ -0,0 +1,197 @@
+        using Env = gameModel::Environment;
+        using Cell = gameModel::Cell;
+        auto players = env.getAllPlayers();
+        ret.reserve(n * n);
+        for(int x = target.x - n / 2; x < target.x + n / 2; x++){
+            for(int y = target.y - n / 2; y < target.y + n / 2; y++){
+                if(gameModel::Position{x, y} == target){
+                    continue;
+                }
+
+                if(Env::getCell(x, y) == Cell::OutOfBounds){
+                    continue;
+                }
+
+                if(env.cellIsFree({x, y})){
+                    ret.emplace_back(x, y);
+                }
+            }
+        }
+
+        return ret;
+    }
+
+
+    // fertig
+    auto Move::check(const gameModel::Environment &envi) const -> ActionResult{
+
+        if (gameModel::Environment::getCell(this->target) == gameModel::Cell::OutOfBounds ||
+            gameController::getDistance(this->actor.get()->position, this->target) > 1) {
+            return ActionResult::Impossible;
+        }
+
+        if (this->checkForFoul(envi) == gameModel::Foul::None) {
+            return ActionResult::Success;
+        }
+        else {
+            return ActionResult::Foul;
+        }
+    }
+
+    auto Move::executeAll(const gameModel::Environment &envi) const ->
+        std::vector<std::pair<gameModel::Environment, double>>{
+
+        std::vector<std::pair<gameModel::Environment, double>> resultVect;
+
+        // @toDo: da fehlt noch alles
+
+        return resultVect;
+    }
+
+    Move::Move(std::shared_ptr<gameModel::Player> actor, gameModel::Position target): Action(actor, target) {}
+
+    void Move::execute(gameModel::Environment &envi) const {
+        ActionResult actionResult = this->check(envi);
+
+        switch (actionResult) {
+
+            case ActionResult::Foul :
+                // @ToDo was passiert bei einem Faul???
+
+            case ActionResult::Success :
+
+                if (envi.getPlayer(this->target) != nullptr) {
+                    movePlayerOnEmptyCell(envi, this->target);
+                }
+                this->actor.get()->position = this->target;
+
+            break;
+
+            case ActionResult::Impossible :
+                throw std::runtime_error("The Selected move is impossible!");
+        }
+    }
+
+    void Move::movePlayerOnEmptyCell(gameModel::Environment &envi, const gameModel::Position &position) const {
+        const std::vector<gameModel::Position> positions = envi.getAllPlayerFreeCellsAround(position);
+        // @ToDo: random cell auswählen & player verschieben
+
+        // init random seed
+        srand (time(NULL));
+
+        int randCellNo = rand() % positions.size();
+
+
+    }
+
+    // fertig
+    auto Move::successProb(const gameModel::Environment&) const -> double {
+        if (gameModel::Environment::getCell(this->target) == gameModel::Cell::OutOfBounds ||
+            gameController::getDistance(this->actor.get()->position, this->target) > 1){
```

Comment:
```suggestion
            gameController::getDistance(this->actor->position, this->target) > 1){
```
wie oben mit shared_ptr